File: Julia/Claude/NBodySimulationPar.jl
module NBodySimulation

using LinearAlgebra
using Random
using Printf
using Base.Threads

"""
    Body

Represents a body in 3D space with position, velocity, and mass.
"""
struct Body
    position::Vector{Float64}  # 3D position vector
    velocity::Vector{Float64}  # 3D velocity vector
    mass::Float64              # Mass of the body
end

"""
    System

Contains all bodies in the simulation.
"""
struct System
    bodies::Vector{Body}
end

"""
    initialize_system(central_mass, central_position, central_velocity, 
                     num_orbiting, orbiting_mass, 
                     min_radius, max_radius)

Initialize a system with a central body and `num_orbiting` smaller bodies
in circular orbits at randomly distributed radii.
"""
function initialize_system(central_mass, central_position, central_velocity, 
                         num_orbiting, orbiting_mass, 
                         min_radius, max_radius)
    # Create central body
    central_body = Body(central_position, central_velocity, central_mass)
    
    # Initialize system with central body
    bodies = [central_body]
    
    # Create orbiting bodies with random positions on circular orbits
    rng = MersenneTwister(1234)  # For reproducibility
    
    for i in 1:num_orbiting
        # Random radius between min_radius and max_radius
        r = min_radius + (max_radius - min_radius) * rand(rng)
        
        # Random point on a sphere
        θ = 2π * rand(rng)  # Azimuthal angle
        φ = acos(2 * rand(rng) - 1)  # Polar angle
        
        # Convert to Cartesian coordinates
        x = r * sin(φ) * cos(θ)
        y = r * sin(φ) * sin(θ)
        z = r * cos(φ)
        
        position = central_position + [x, y, z]
        
        # Calculate velocity for a circular orbit
        # v = sqrt(G * M / r)
        G = 6.67430e-11  # Gravitational constant
        v_orbit = sqrt(G * central_mass / r)
        
        # Velocity vector perpendicular to position vector (for circular orbit)
        # We need a vector perpendicular to the radial vector
        if abs(z) < 0.9 * r  # Avoid numerical issues near poles
            perp_vector = cross([0.0, 0.0, 1.0], [x, y, z])
        else
            perp_vector = cross([1.0, 0.0, 0.0], [x, y, z])
        end
        
        # Normalize and scale to orbital velocity
        velocity_direction = perp_vector / norm(perp_vector)
        velocity = central_velocity + velocity_direction * v_orbit
        
        # Create and add the orbiting body
        push!(bodies, Body(position, velocity, orbiting_mass))
    end
    
    return System(bodies)
end

"""
    compute_acceleration(system, body_index)

Compute the acceleration of a body due to gravitational forces from all other bodies.
"""
function compute_acceleration(system, body_index)
    G = 6.67430e-11  # Gravitational constant
    ε = 1e-10        # Softening parameter to avoid division by zero
    
    body = system.bodies[body_index]
    acceleration = zeros(Float64, 3)
    
    for i in 1:length(system.bodies)
        if i == body_index
            continue  # Skip self-interaction
        end
        
        other = system.bodies[i]
        r_vec = other.position - body.position
        r_squared = sum(r_vec.^2) + ε^2  # Softened distance squared
        r = sqrt(r_squared)
        
        # Gravitational acceleration: a = G * m / r^2 * (r_vec / r)
        acceleration += G * other.mass / r_squared * (r_vec / r)
    end
    
    return acceleration
end

"""
    kick_step!(system, dt)

Update velocities of all bodies using computed accelerations (kick step).
Uses multithreading for computing accelerations.
"""
function kick_step!(system, dt)
    n = length(system.bodies)
    accelerations = Vector{Vector{Float64}}(undef, n)
    
    # Compute accelerations in parallel
    @threads for i in 1:n
        accelerations[i] = compute_acceleration(system, i)
    end
    
    # Update velocities
    for i in 1:n
        system.bodies[i] = Body(
            system.bodies[i].position,
            system.bodies[i].velocity + accelerations[i] * dt,
            system.bodies[i].mass
        )
    end
end

"""
    drift_step!(system, dt)

Update positions of all bodies using their velocities (drift step).
Uses multithreading for updating positions.
"""
function drift_step!(system, dt)
    n = length(system.bodies)
    new_bodies = Vector{Body}(undef, n)
    
    @threads for i in 1:n
        body = system.bodies[i]
        new_bodies[i] = Body(
            body.position + body.velocity * dt,
            body.velocity,
            body.mass
        )
    end
    
    # Copy the new bodies back to the system
    for i in 1:n
        system.bodies[i] = new_bodies[i]
    end
end

"""
    calculate_energy(system)

Calculate the total energy (kinetic + potential) of the system.
Uses multithreading for kinetic energy calculation and atomic operations
for thread-safe summation of potential energy.
"""
function calculate_energy(system)
    G = 6.67430e-11  # Gravitational constant
    ε = 1e-10        # Softening parameter
    
    n = length(system.bodies)
    
    # Calculate kinetic energy in parallel: KE = 0.5 * m * v^2
    kinetic_energies = zeros(Float64, nthreads())
    
    @threads for i in 1:n
        thread_id = threadid()
        body = system.bodies[i]
        v_squared = sum(body.velocity.^2)
        kinetic_energies[thread_id] += 0.5 * body.mass * v_squared
    end
    
    kinetic_energy = sum(kinetic_energies)
    
    # For potential energy, we need to be careful with parallelization
    # because of the double loop. We'll use atomic operations for thread safety.
    potential_energy_chunks = zeros(Float64, nthreads())
    
    # Each thread handles a chunk of the outer loop
    @threads for i in 1:(n-1)
        thread_id = threadid()
        local_potential = 0.0
        
        for j in (i+1):n
            body_i = system.bodies[i]
            body_j = system.bodies[j]
            
            r_vec = body_j.position - body_i.position
            r = sqrt(sum(r_vec.^2) + ε^2)  # Softened distance
            
            local_potential -= G * body_i.mass * body_j.mass / r
        end
        
        potential_energy_chunks[thread_id] += local_potential
    end
    
    potential_energy = sum(potential_energy_chunks)
    
    return kinetic_energy + potential_energy
end

"""
    run_simulation!(system, dt, num_steps)

Run the N-body simulation for a specified number of steps.
Reports thread usage and performance metrics.
"""
function run_simulation!(system, dt, num_steps)
    println("Running simulation with $(nthreads()) threads")
    
    initial_energy = calculate_energy(system)
    println("Initial total energy: ", initial_energy)
    
    # Start timing
    start_time = time()
    
    # Run the simulation for num_steps
    for step in 1:num_steps
        if step % (num_steps ÷ 10) == 0
            elapsed_so_far = time() - start_time
            steps_per_second = step / elapsed_so_far
            estimated_total = num_steps / steps_per_second
            
            println("Completed step $step of $num_steps ($(round(100.0 * step / num_steps, digits=1))%)")
            println("Current performance: $(round(steps_per_second, digits=2)) steps/second")
            println("Estimated total runtime: $(round(estimated_total / 60, digits=1)) minutes")
        end
        
        # Kick-drift-kick leapfrog integrator
        kick_step!(system, dt/2)  # Half kick
        drift_step!(system, dt)   # Full drift
        kick_step!(system, dt/2)  # Half kick
    end
    
    # End timing
    end_time = time()
    elapsed = end_time - start_time
    
    final_energy = calculate_energy(system)
    println("Final total energy: ", final_energy)
    println("Energy change: ", final_energy - initial_energy)
    println("Relative energy change: ", (final_energy - initial_energy) / initial_energy)
    println("Simulation completed in $(elapsed) seconds")
    println("Average performance: $(round(num_steps / elapsed, digits=2)) steps/second")
    
    return system
end

"""
    main(num_orbiting)

Create and run an N-body simulation with a central body and a specified number of orbiting bodies.
Shows thread information and performance metrics.
"""
function main(num_orbiting=10000)
    # Display thread information
    println("Julia is running with $(nthreads()) threads")
    println("To increase threads, set the JULIA_NUM_THREADS environment variable")
    println("or start Julia with the --threads=auto or --threads=N flag")
    
    # Parameters
    central_mass = 1.0e30      # Mass of central body (approximately solar mass)
    central_position = [0.0, 0.0, 0.0]
    central_velocity = [0.0, 0.0, 0.0]
    orbiting_mass = 1.0e20     # Mass of orbiting bodies
    min_radius = 1.0e10        # Minimum orbital radius
    max_radius = 5.0e10        # Maximum orbital radius
    dt = 3600.0                # Time step (seconds)
    num_steps = 100           # Number of simulation steps
    
    println("Initializing system with $num_orbiting orbiting bodies...")
    init_start = time()
    
    # Initialize the system
    system = initialize_system(
        central_mass, central_position, central_velocity,
        num_orbiting, orbiting_mass, min_radius, max_radius
    )
    
    init_time = time() - init_start
    println("Initialization completed in $(round(init_time, digits=2)) seconds")
    println("Starting simulation with $(length(system.bodies)) bodies for $num_steps steps...")
    
    # Run the simulation
    run_simulation!(system, dt, num_steps)
    
    println("Simulation complete!")
    return system
end

"""
    reduced_main(num_orbiting=1000)

Run the simulation with a smaller number of particles for testing.
"""
function reduced_main(num_orbiting=10000)
    main(num_orbiting)
end

end  # module NBodySimulation

# Execute the simulation with reduced particle count for testing
# Comment out the next line to run the full simulation
#NBodySimulation.reduced_main()

# Uncomment the next line to run the full simulation with 1 million bodies
NBodySimulation.main()

File: Java/Claude/NBodySimulationPar.java
import java.util.Random;
import java.util.Arrays;
import java.util.stream.IntStream;

public class NBodySimulationPar {
    // Gravitational constant (using an arbitrary value for simulation purposes)
    private static final double G = 6.67430e-11;
    
    // Simulation bodies
    private Body[] bodies;
    
    // Simulation parameters
    private double timeStep;
    private int numSteps;
    
    // Class to represent a body in the simulation
    static class Body {
        // Position components
        double x, y, z;
        // Velocity components
        double vx, vy, vz;
        // Mass
        double mass;
        
        public Body(double x, double y, double z, double vx, double vy, double vz, double mass) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx;
            this.vy = vy;
            this.vz = vz;
            this.mass = mass;
        }
        
        // Copy constructor
        public Body(Body other) {
            this.x = other.x;
            this.y = other.y;
            this.z = other.z;
            this.vx = other.vx;
            this.vy = other.vy;
            this.vz = other.vz;
            this.mass = other.mass;
        }
    }
    
    public NBodySimulationPar(int numBodies, double timeStep, int numSteps) {
        this.bodies = new Body[numBodies];
        this.timeStep = timeStep;
        this.numSteps = numSteps;
    }
    
    /**
     * Initialize a central body with specified number of smaller bodies on circular orbits
     * @param centralMass Mass of the central body
     * @param bodyMass Mass of each orbiting body
     * @param minRadius Minimum orbit radius
     * @param maxRadius Maximum orbit radius
     */
    public void initializeSystem(double centralMass, double bodyMass, double minRadius, double maxRadius) {
        // Central body at origin with no initial velocity
        bodies[0] = new Body(0, 0, 0, 0, 0, 0, centralMass);
        
        // Create a thread-safe random number generator with fixed seed for reproducibility
        Random seedGenerator = new Random(42);
        long seed = seedGenerator.nextLong();
        
        // Use parallel streams to initialize the orbiting bodies
        IntStream.range(1, bodies.length)
            .parallel()
            .forEach(i -> {
                // Each thread needs its own random generator to avoid contention
                // Use a different seed for each thread based on the index
                Random random = new Random(seed + i);
                
                // Random radius between minRadius and maxRadius
                double radius = minRadius + random.nextDouble() * (maxRadius - minRadius);
                
                // Random position on a sphere of the given radius
                double theta = random.nextDouble() * 2 * Math.PI;
                double phi = Math.acos(2 * random.nextDouble() - 1);
                
                double x = radius * Math.sin(phi) * Math.cos(theta);
                double y = radius * Math.sin(phi) * Math.sin(theta);
                double z = radius * Math.cos(phi);
                
                // Calculate velocity for a circular orbit
                // v = sqrt(G * M / r)
                double velocity = Math.sqrt(G * centralMass / radius);
                
                // Direction perpendicular to radius vector (for circular orbit)
                // We need a vector perpendicular to (x, y, z)
                // One approach is to cross product with a reference vector
                double[] perpDirection = crossProduct(x, y, z, 0, 0, 1);
                
                // If the result is close to zero, use a different reference vector
                if (magnitude(perpDirection) < 1e-10) {
                    perpDirection = crossProduct(x, y, z, 0, 1, 0);
                }
                
                // Normalize the perpendicular direction
                double perpMagnitude = magnitude(perpDirection);
                perpDirection[0] /= perpMagnitude;
                perpDirection[1] /= perpMagnitude;
                perpDirection[2] /= perpMagnitude;
                
                // Set velocity components
                double vx = velocity * perpDirection[0];
                double vy = velocity * perpDirection[1];
                double vz = velocity * perpDirection[2];
                
                bodies[i] = new Body(x, y, z, vx, vy, vz, bodyMass);
            });
    }
    
    // Helper function to compute cross product
    private double[] crossProduct(double x1, double y1, double z1, double x2, double y2, double z2) {
        double[] result = new double[3];
        result[0] = y1 * z2 - z1 * y2;
        result[1] = z1 * x2 - x1 * z2;
        result[2] = x1 * y2 - y1 * x2;
        return result;
    }
    
    // Helper function to compute magnitude of a vector
    private double magnitude(double[] vector) {
        return Math.sqrt(vector[0] * vector[0] + vector[1] * vector[1] + vector[2] * vector[2]);
    }
    
    // Calculate total energy (kinetic + potential) of the system
    public double calculateTotalEnergy() {
        // Calculate kinetic energy for each body using parallel streams
        double kineticEnergy = Arrays.stream(bodies)
            .parallel()
            .mapToDouble(body -> {
                double speedSquared = body.vx * body.vx + body.vy * body.vy + body.vz * body.vz;
                return 0.5 * body.mass * speedSquared;
            })
            .sum();
        
        // Calculate potential energy using parallel streams
        double potentialEnergy = IntStream.range(0, bodies.length)
            .parallel()
            .mapToDouble(i -> {
                Body bi = bodies[i];
                return IntStream.range(i + 1, bodies.length)
                    .mapToDouble(j -> {
                        Body bj = bodies[j];
                        
                        double dx = bi.x - bj.x;
                        double dy = bi.y - bj.y;
                        double dz = bi.z - bj.z;
                        
                        double distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
                        
                        // Avoid division by zero
                        if (distance > 0) {
                            return -G * bi.mass * bj.mass / distance;
                        }
                        return 0.0;
                    })
                    .sum();
            })
            .sum();
            
        return kineticEnergy + potentialEnergy;
    }
    
    // Run the simulation using a first-order kick-step method (leapfrog)
    public void runSimulation() {
        // Display initial energy
        double initialEnergy = calculateTotalEnergy();
        System.out.println("Initial total energy: " + initialEnergy);
        
        // Run simulation steps
        for (int step = 0; step < numSteps; step++) {
            if (step % 100 == 0) {
                System.out.println("Step " + step + " of " + numSteps);
            }
            
            updateVelocities();
            updatePositions();
        }
        
        // Display final energy
        double finalEnergy = calculateTotalEnergy();
        System.out.println("Final total energy: " + finalEnergy);
        System.out.println("Energy change: " + (finalEnergy - initialEnergy));
        System.out.println("Relative energy change: " + ((finalEnergy - initialEnergy) / initialEnergy));
    }
    
    // Update velocities using forces (kick step)
    private void updateVelocities() {
        // Use parallel streams for multithreading
        java.util.stream.IntStream.range(0, bodies.length)
            .parallel()
            .forEach(i -> {
                Body bi = bodies[i];
                double fx = 0, fy = 0, fz = 0;
                
                // Calculate net force on body i from all other bodies
                for (int j = 0; j < bodies.length; j++) {
                    if (i == j) continue;
                    
                    Body bj = bodies[j];
                    
                    double dx = bj.x - bi.x;
                    double dy = bj.y - bi.y;
                    double dz = bj.z - bi.z;
                    
                    double distanceSquared = dx * dx + dy * dy + dz * dz;
                    double distance = Math.sqrt(distanceSquared);
                    
                    // Add small softening parameter to avoid numerical instability
                    double softening = 1e-10;
                    distanceSquared += softening;
                    
                    // Calculate gravitational force magnitude (F = G * m1 * m2 / r^2)
                    double forceMagnitude = G * bi.mass * bj.mass / distanceSquared;
                    
                    // Force components (F * direction)
                    fx += forceMagnitude * dx / distance;
                    fy += forceMagnitude * dy / distance;
                    fz += forceMagnitude * dz / distance;
                }
                
                // Update velocity using F = ma => a = F/m => Δv = a * Δt
                bi.vx += fx / bi.mass * timeStep;
                bi.vy += fy / bi.mass * timeStep;
                bi.vz += fz / bi.mass * timeStep;
            });
    }
    
    // Update positions using velocities (drift step)
    private void updatePositions() {
        // Use parallel streams for updating positions
        java.util.Arrays.stream(bodies)
            .parallel()
            .forEach(body -> {
                body.x += body.vx * timeStep;
                body.y += body.vy * timeStep;
                body.z += body.vz * timeStep;
            });
    }
    
    // Main method to run the simulation
    public static void main(String[] args) {
        // Number of bodies (1 central + orbiting bodies)
        final int NUM_BODIES = 10_001; // 1 million small bodies + 1 central body
        
        // Time step (smaller is more accurate but slower)
        final double TIME_STEP = 0.001 * 3600*24*365;
        
        // Number of simulation steps
        final int NUM_STEPS = 100;
        
        // Create simulation
        NBodySimulationPar simulation = new NBodySimulationPar(NUM_BODIES, TIME_STEP, NUM_STEPS);
        
        // Central mass (e.g., like a star)
        double centralMass = 1.0e30;
        
        // Orbiting body mass (e.g., like planets)
        double bodyMass = 1.0e20;
        
        // Orbit radius range
        double minRadius = 1.0e8;
        double maxRadius = 5.0e8;
        
        System.out.println("Initializing " + (NUM_BODIES - 1) + " bodies around central mass...");
        long startTime = System.currentTimeMillis();
        
        // Initialize the system
        simulation.initializeSystem(centralMass, bodyMass, minRadius, maxRadius);
        
        long initTime = System.currentTimeMillis() - startTime;
        System.out.println("Initialization completed in " + initTime + " ms");
        
        // Run the simulation
        System.out.println("Running simulation for " + NUM_STEPS + " steps...");
        startTime = System.currentTimeMillis();
        
        simulation.runSimulation();
        
        long simTime = System.currentTimeMillis() - startTime;
        System.out.println("Simulation completed in " + simTime + " ms");
    }
}